### ISSUE TITLE
Intrinsic inquiry functions

### ISSUE THREAD

**Issue by gxyd** [lfortran/lfortran#3688]
From 18-007r1 documentation:

> <b>3.89 inquiry function</b>
intrinsic function, or function in an intrinsic module, whose result depends on the properties of one or more of its arguments instead of their values

Currently, intrinsic functions like `kind`, `len` are implemented as "Elemental intrinsic functions" & behave in odd manner. For e.g.

```f90
program main
    print *, kind([1, 2, 3]) !< LFortran outputs this as 4, 4, 4 instead of 4
end program
```

similarly:
```f90
program main
    character(len=6) :: x(2)
    print *, len(x)
end program
```

on executing the above program results in an error code `11` as output.

There are other intrinsic functions which currently implemented in LFortran as elementary intrinsic functions, which actually are intrinsic inquiry functions: https://www.ibm.com/docs/en/xl-fortran-linux/16.1.0?topic=procedures-inquiry-intrinsic-functions

**Comment by gxyd** [comment]
I'm trying to work on this issue now.

**Comment by gxyd** [comment]
@certik , let me know what you think of this issue. And if it's actually that important for us

**Comment by Thirumalai-Shaktivel** [comment]
I think `kind` has to be represented as `TypeInquiry`, as we can compute the output during compile time.
For the other example, the array needs to be initialised, so the seg fault.

**Comment by gxyd** [comment]
> I think kind has to be represented as TypeInquiry, as we can compute the output during compile time.

Is that something we already have in our codebase?

> For the other example, the array needs to be initialised, so the seg fault.

I don't fully agree to it, as I think we should still be able to give "inquiry functions", since it's definition says:
> whose result depends on the properties of one or more of its arguments instead of their values

As it's just dependent on it's properties, so I don't think we necessarily need the array to be initialised to give output of "inquiry functions"

**Comment by Thirumalai-Shaktivel** [comment]
Yes, see: https://github.com/lfortran/lfortran/pull/3309,  https://docs.lfortran.org/en/asr/asr_nodes/expression_nodes/TypeInquiry/

String are not handled correctly in the backend (it leaks memory sometimes, has bugs, ...), so the issue. We will fix them soon.

**Comment by gxyd** [comment]
> Yes, see: https://github.com/lfortran/lfortran/pull/3309, https://docs.lfortran.org/en/asr/asr_nodes/expression_nodes/TypeInquiry/

That looks good, I should've searched for it maybe before submitting the issue I guess. Thanks for this @Thirumalai-Shaktivel
